core/src/main/java/org/infinispan/distribution/LocalizedCacheTopology.java

```diff
@@ -0,0 +1,190 @@
+package org.infinispan.distribution;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import org.infinispan.commons.hash.MurmurHash3;
+import org.infinispan.commons.util.Immutables;
+import org.infinispan.commons.util.IntSet;
+import org.infinispan.configuration.cache.CacheMode;
+import org.infinispan.distribution.ch.ConsistentHash;
+import org.infinispan.distribution.ch.KeyPartitioner;
+import org.infinispan.distribution.ch.impl.ReplicatedConsistentHash;
+import org.infinispan.remoting.transport.Address;
+import org.infinispan.topology.CacheTopology;
+
+/**
+ * Extends {@link CacheTopology} with information about keys owned by the local node.
+ *
+ * @author Dan Berindei
+ * @since 9.0
+ */
+public class LocalizedCacheTopology extends CacheTopology {
+
+   private final Address localAddress;
+   private final KeyPartitioner keyPartitioner;
+   private final boolean isDistributed;
+   private final boolean isReplicated;
+   private final int numSegments;
+   private final int numOwners;
+   private final DistributionInfo[] distributionInfos;
+
+   public static LocalizedCacheTopology makeSingletonTopology(CacheMode cacheMode, Address localAddress) {
+      List<Address> members = Collections.singletonList(localAddress);
+      ConsistentHash ch = new ReplicatedConsistentHash(MurmurHash3.getInstance(), members, new int[]{0});
+      CacheTopology cacheTopology = new CacheTopology(0, 0, ch, null, members, null);
+      return new LocalizedCacheTopology(cacheMode, cacheTopology, key -> 0, localAddress);
+   }
+
+   public LocalizedCacheTopology(CacheMode cacheMode, CacheTopology cacheTopology, KeyPartitioner keyPartitioner,
+                                 Address localAddress) {
+      super(cacheTopology.getTopologyId(), cacheTopology.getRebalanceId(), cacheTopology.getCurrentCH(),
+            cacheTopology.getPendingCH(), cacheTopology.getUnionCH(), cacheTopology.getActualMembers(),
+            cacheTopology.getMembersPersistentUUIDs());
+
+      ConsistentHash readCH = getReadConsistentHash();
+      ConsistentHash writeCH = getWriteConsistentHash();
+
+      this.localAddress = localAddress;
+      this.keyPartitioner = keyPartitioner;
+      this.isDistributed = cacheMode.isDistributed();
+      this.isReplicated = cacheMode.isReplicated();
+      this.numSegments = readCH.getNumSegments();
+      this.numOwners = readCH.getNumOwners();
+
+      if (isDistributed) {
```

Comment:
This happens when the topology changes, and I wanted to avoid having `StateConsumerImpl` depend on `ClusteringDependentLogic`. Besides, I have this feeling that `ClusteringDependentLogic` is doing way too much, I'd love to use `LocalizedCacheTopology` (or maybe some subclasses) for all the ownership stuff and to leave `ClusteringDependentLogic` dealing only with the commit logic (at which point we could also rename it to something more specific).